lib/active_support/cache/dalli_store.rb

```diff
@@ -141,7 +135,17 @@ def write(name, value, options=nil)
         instrument(:write, name, options) do |payload|
           with do |connection|
             options = options.merge(:connection => connection)
-            write_entry(name, value, options)
+            # Support :race_condition_ttl by appending a secondary expiration to the value.
+            if options[:race_condition_ttl] &&
+              (ttl = options[:expires_in] || @options[:expires_in])
+              options[:expires_in] = ttl + options[:race_condition_ttl] * 2
+              entry = Entry.new
```

Comment:
@wjordan This method now feels a lot longer and more complex.  I'd like to see it broken up a bit.  For example, the three lines generating the entry could easily be pulled out into a class method on the Entry class.  Similarly, the logic for the expires_in and ttl is difficult to follow as it's wrapped in the if condition.